src/arbitrary/_std/heap.rs

```diff
@@ -0,0 +1,58 @@
+//-
+// Copyright 2017, 2018 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Arbitrary implementations for `std::hash`.
+
+use std::heap::*;
+use std::usize;
+
+use strategy::{Strategy, BoxedStrategy, Just, TupleUnion, W};
+use strategy::statics::static_map;
+use arbitrary::{any, SMapped};
+
+arbitrary!(CannotReallocInPlace; CannotReallocInPlace);
+arbitrary!(Heap; Heap);
+
+// Not Debug.
+//lazy_just!(System, || System);
+
+arbitrary!(Layout, BoxedStrategy<Layout>;
+    (0u8..32u8).prop_flat_map(|align_power| {
+        // align must be a power of two and <= (1 << 31):
+        let align = 1 << align_power;
+        // Compute the highest multiple of align <= usize::MAX:
+        // By definition an integer since 2^X / 2^Y = 2^(X - Y)
+        // X, Y are integers, X >= Y, so X - Y is a positive integer,
+        // so 2^(X - Y) is too.
+        let max_size = usize::MAX / align;
+        // Should perhaps be ..=max_size but we can't express that now.
+        (..max_size).prop_map(move |size|
```

Comment:
This isn't quite correct (e.g., the code here says that a 32-bit architecture can't make an allocation larger than 2 bytes with alignment of 2³¹), and can be rewritten without the flat_map anyway:

```rust
(0u8..32u8, prop::usize::ANY).prop_map(|align_power, size| {
    let align = 1usize << align_power;
    let max_size = 0usize.wrapping_sub(align);
    // Not quite a uniform distribution due to clamping, but probably good enough
    Layout::from_size_align(align, std::cmp::min(max_size, size)).unwrap()
})
```

Reply:
That works well, changed to it =)